### PR TITLE
feature: add opt-in Turso storage backend

### LINKED ISSUE #420: Verify Turso support, and potentially tweak it
Verify we can work with Turso, and potentially tweak our adapter to work with Turso.

We support remtoe connections to sqllite DBs, so it will most likely boil down to handling their specific approach if it doesn't work already

### PR DESCRIPTION
**Pull request by jamby77**
Closes #420.

## What this adds

An **opt-in** Turso / libSQL storage backend for the API, alongside the existing SQLite, PostgreSQL and Memory options.

- `apps/api/src/storage/adapters/libsql-driver.ts` — libSQL driver implementing the same driver surface as the local `better-sqlite3` path
- `apps/api/src/storage/adapters/sqlite.adapter.ts` — selects the remote driver when a libSQL URL is configured, local file otherwise
- `apps/api/src/storage/factory/storage-client.factory.ts` — wiring for the new backend
- `apps/api/src/config/env.schema.ts` — Zod validation for the new Turso env vars
- `docs/configuration.md`, `docs/troubleshooting.md`, `.env.example`, `README.md` — setup and failure-mode docs

Existing deployments are unaffected: with no Turso URL configured the adapter takes the same local-file path as before.

## Verification

| Check | Result |
| --- | --- |
| `libsql-driver.spec.ts` (unit) | 18/18 pass |
| `turso-storage.e2e-spec.ts` vs local `libsql-server` over HTTP | 8/8 pass |
| `turso-storage.e2e-spec.ts` vs hosted Turso (`libsql://`, TLS + auth token) | 8/8 pass |
| Full API suite | 3098 pass, 1 pre-existing `license.service` failure unrelated to this branch |

The e2e run against hosted Turso used a throwaway database, since the spec runs migrations and seeds the settings row; it was destroyed afterwards.

The e2e suite is gated on `TURSO_TEST_URL` (`describeTurso = TURSO_URL ? describe : describe.skip`), so it stays skipped in CI unless credentials are provided. Coverage includes schema creation and migrations over the network, batched transactions for ACL entries / client snapshots / anomaly events, upsert refresh, rollback on a failed transaction, and settings round-trip parity with local SQLite.

## Second commit — unrelated test-infra fix

`bugfix: run valkey cluster-init as a single shell command` fixes a pre-existing bug in `docker-compose.test.yml`, last touched in #108.

The `valkey-cluster-init` entrypoint used a folded scalar (`>`) whose argument lines were indented deeper than the first content line. YAML preserves newlines on more-indented lines in folded style, so the block reached the shell as five separate commands rather than one:

```
[ERR] Wrong number of arguments for specified --cluster sub command
sh: 127.0.0.1:6401: not found
```

The container exited 127, which made `docker compose up -d` fail, which `apps/api/test/global-setup.ts` runs via `execSync` — so Jest aborted during global setup before loading a single test file. With the fix the container exits 0 with `All 16384 slots covered.`

Worth a look from someone who knows the CI setup: either the cluster suites have been silently unexercised, or CI never brings that service up.

---

> [!NOTE]
> **Medium Risk**
> Changes core persistence initialization and remote transaction semantics; misconfiguration or driver bugs could affect audit/analytics data integrity, though the backend is opt-in and local sqlite defaults are unchanged.
> 
> **Overview**
> Adds an **opt-in** fourth persistence backend (`STORAGE_TYPE=turso`) so Docker and serverless deployments can use SQLite semantics over libSQL without the `better-sqlite3` native module.
> 
> **Configuration & wiring:** `STORAGE_URL` and optional `STORAGE_AUTH_TOKEN` are validated at startup (`libsql://` requires a token; `http://` rejects a token to block cleartext credentials). The storage factory routes `turso` through the existing `SqliteAdapter` with `{ url, authToken }` instead of a local filepath.
> 
> **Drivers:** Remote connections use the new `libsql` package plus a compatibility layer that re-prepares statements across transaction boundaries and emulates nested `transaction()` with savepoints—behavior remote libSQL needs but local SQLite did not. Local `STORAGE_TYPE=sqlite` still uses dynamically loaded `better-sqlite3`, with clearer errors when the native module is missing.
> 
> **Packaging & docs:** `.dockerignore` no longer drops the SQLite adapter source (only adapter tests); README/docs/`.env.example` document Turso setup and auth edge cases. Optional e2e tests run when `TURSO_TEST_URL` is set.
> 
> **Incidental:** `docker-compose.test.yml` fixes Valkey cluster init so `valkey-cli --cluster create` runs as one shell command.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ae82d01c1ca7a178c672a6a3078b86330c66ce01. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added Turso/libSQL support for remote SQLite databases.
  * Added storage URL and authentication-token configuration with validation.
  * Added Docker compatibility guidance and Turso usage examples.

* **Bug Fixes**
  * Improved local SQLite driver loading and error messages when native support is unavailable.
  * Enhanced transaction handling, statement reuse, and rollback behavior for remote SQLite connections.

* **Documentation**
  * Updated configuration, troubleshooting, and setup documentation for Turso and SQLite requirements.

### CODE REVIEW

File: apps/api/src/storage/adapters/__tests__/libsql-driver.spec.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+      }
+      originalExec(sql);
+    });
+
+    const failing = db.transaction(() => {
+      throw new Error('constraint failed');
+    });
+
+    expect(failing).toThrow('constraint failed');
+  });
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/api/src/storage/adapters/__tests__/libsql-driver.spec.ts:299]: Fixed in d30d70d2.

Confirmed: `patchForRemote` captures `db.exec.bind(db)` at patch time, so the spy installed after `openRemote()` was never reached and the simulated ROLLBACK failure never happened — the test passed on the unrelated `constraint failed` throw.

The failure now originates inside `FakeDatabase` via a `failExec` flag, which the captured `nativeExec` does call. Added `expect(log).toEqual(['exec:BEGIN', 'exec:ROLLBACK'])` so the rollback path has to have executed for the test to pass.

File: apps/api/src/storage/adapters/libsql-driver.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        }
+        throw error;
+      } finally {
+        depth = 0;
+        generation += 1;
+      }
+    };
+
+    return wrapped as unknown as F;
+  };
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/api/src/storage/adapters/libsql-driver.ts:172]: Fixed in d30d70d2.

A nested body now runs inside `SAVEPOINT betterdb_sp_<depth>`, releasing on success and doing `ROLLBACK TO` + `RELEASE` on throw, so a nested write no longer rides on the outer COMMIT when the outer body swallows the error.

Added a test asserting the full exec sequence for exactly that case (outer insert, savepoint, inner insert, rollback to, release, commit).

Also verified `SAVEPOINT` / `ROLLBACK TO` / `RELEASE` are accepted over the remote protocol rather than only in local SQLite — probed against a `libsql-server` instance over HTTP, where the nested row was unwound and the outer row survived the commit.

File: apps/api/test/turso-storage.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+  afterAll(async () => {
+    if (storage) {
+      await storage.close();
+    }
+  });
+
+  it('creates the schema and runs migrations over the network', async () => {
+    const settings = await storage.getSettings();
+    expect(settings).toBeDefined();
+  });
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/api/test/turso-storage.e2e-spec.ts:127]: Fixed in d30d70d2, though not quite as suggested — `not.toBeNull()` would fail here.

`initialize()` does not seed a default settings row: the later test `rejects an update when no settings row exists` depends on that, and there is no `INSERT INTO app_settings` on the migration path. So `null` is the correct value for a fresh database, and the original `toBeDefined()` was weak for exactly the reason you describe.

Changed to `await expect(storage.getSettings()).resolves.toBeNull()`. A missing table rejects rather than resolving, so this still fails if the schema step silently did not run, which is what the test name claims to cover. Verified against a live libSQL server.

File: apps/api/src/storage/adapters/libsql-driver.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  return new Proxy(target, {
+    get(_target, property: string | symbol): unknown {
+      const current = ensure() as unknown as StatementRecord;
+      const value = current[property as string];
+      if (typeof value === 'function') {
+        return (value as (...args: unknown[]) => unknown).bind(current);
+      }
+      return value;
+    },
+  }) as unknown as Database.Statement;
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/api/src/storage/adapters/libsql-driver.ts]: Not taking this one — it describes capability the codebase does not use.

Grepped `apps/api/src` and `packages/shared/src`: there are zero calls to `.pluck()`, `.expand()`, `.raw()`, or `.safeIntegers()`, and no `.bind()` on a prepared statement. Every call site goes straight from `db.prepare(sql)` to `.run()` / `.get()` / `.all()`. This driver is internal to `SqliteAdapter`, not a published API, so there is no external caller to account for either.

The underlying observation is correct — the fluent methods return the native statement, so a chained call would escape the generation check and could use a stale statement across a transaction boundary, and `ensure()` would drop the configuration on re-preparation. Record-and-replay plus returning the proxy from the fluent methods is a real amount of machinery to carry for a path nothing currently takes, so I would rather add it alongside the first caller that needs it, when the required semantics are concrete.

Resolving as won't-fix for now rather than leaving it hanging; happy to reopen if a use case lands.

File: apps/api/src/storage/adapters/libsql-driver.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+          nativeExec(`RELEASE ${savepoint}`);
+          return result;
+        } catch (error) {
+          try {
+            nativeExec(`ROLLBACK TO ${savepoint}`);
+            nativeExec(`RELEASE ${savepoint}`);
+          } catch {
+            // The server already unwound past this savepoint; surface the
+            // original error.
+          }
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/api/src/storage/adapters/libsql-driver.ts:130]: Confirmed and fixed in 2d56846.

The suppressed cleanup failure did let the outer wrapper COMMIT writes that never
unwound: the nested error propagates, the outer body can swallow it, and the outer
path then ran COMMIT unconditionally.

The transaction is now marked rollback-only when `ROLLBACK TO` throws. The outer
wrapper checks that flag before committing, issues ROLLBACK instead, and throws
rather than reporting success. Covered by a new spec case that fails
`ROLLBACK TO betterdb_sp_1` while the outer body swallows the inner error and
asserts ROLLBACK ran and COMMIT did not.

Verified: 20/20 driver unit tests, plus 8/8 storage e2e against a real remote
libSQL server over HTTP.

**Review by KIvanow — CHANGES_REQUESTED** [review]
- The turso case in the factory passes STORAGE_SQLITE_FILEPATH to the adapter, but openDatabase() ignores it once url is set - drop it or add a comment, it reads as if the local file matters.
 - We've got no CI coverage of the actual driver - unit tests mock libsql and the e2e is skipped without TURSO_TEST_URL. Fine to land as-is since it's opt-in, but can we add a file:-backed libSQL run in CI as a follow-up? Cheap, and it'd exercise real prepare/transaction.
 - Minor: https:// passes env validation without an auth token (fails only at first query), and the lazy-statement Proxy only traps get, so .pluck()/.raw() would bypass the re-prepare - nothing hits it today, worth a one-line note.

**Comment by jamby77** [comment]
Addressed in `17b743b7`.

**1. `STORAGE_SQLITE_FILEPATH` on the turso path.** Dropped rather than commented, and the config type now makes the pairing impossible to write:

```ts
export type SqliteAdapterConfig =
  | { filepath: string; url?: undefined; authToken?: undefined }
  | { url: string; authToken?: string; filepath?: undefined };
```

The factory's turso case passes `{ url, authToken }` and no longer reads the filepath var at all; `openDatabase()` branches on `this.config.url !== undefined`. The e2e fixture lost its `filepath: './data/unused-for-remote.db'` too — that was the same misreading, in test form.

**2. No CI coverage of the real driver.** Agreed, and taking it as a follow-up rather than folding it in here — a `file:`-backed libSQL job is a CI change, not an adapter change, and it belongs with the per-package suites work rather than bolted onto this branch. Filing it against #420.

**3a. `https://` without an auth token.** Documented, not hardened. `env.schema.ts` keys its auth-token requirement on `libsql://` specifically, and widening that to `https://` would reject a self-hosted `sqld` behind TLS with authentication off, which is a legitimate deployment. So the gap is real but not a validation error. It is now called out in `.env.example`, `docs/configuration.md`, and a new `docs/troubleshooting.md` section, "Turso starts, then every query fails to authenticate" — the failure mode is a clean startup followed by every query failing, which is exactly the shape that sends someone looking at the wrong layer.

**3b. The `get`-only Proxy.** Fixed rather than noted. The note would have documented a latent bug: `pluck()`/`raw()`/`expand()`/`safeIntegers()`/`bind()` return `this`, so the raw statement escaped the proxy *and* the modifier was silently discarded by the next re-prepare. A caller who chained one across a transaction boundary would have gotten shaped rows back instead of plucked scalars, with no error. `createLazyStatement` now records a modifier call when the return value is the statement itself, replays the recorded modifiers onto every re-prepare, and returns the proxy so the chain stays wrapped.

Mutation-tested: reverting `libsql-driver.ts` fails two of the three new tests (the third is a baseline that passes either way). The re-prepare test asserts the exact driver call log, so a replay that silently stops happening is a failure, not a pass.

23/23 in `libsql-driver.spec.ts`.

File: apps/api/src/storage/factory/storage-client.factory.ts [hunk 6]
```diff
@@ -15,21 +15,22 @@ export class StorageClientFactory {
...
+        break;
+      }
+      case 'turso': {
+        const url = this.configService.get<string>('STORAGE_URL');
+        if (!url) {
+          throw new Error('STORAGE_URL is required for Turso storage');
         }
+        const { SqliteAdapter } = await import('../adapters/sqlite.adapter');
+        const authToken = this.configService.get<string>('STORAGE_AUTH_TOKEN');
+        client = new SqliteAdapter({ url, authToken });
```
[the remark this answers is not in the document]
Comment (jamby77) [reply apps/api/src/storage/factory/storage-client.factory.ts:33]: Valid, fixed in `ae82d01c` — but in `env.schema.ts` rather than the factory, since that is where every other `STORAGE_URL` rule already lives and a rejected boot with a named path beats a throw from the factory.

```
STORAGE_AUTH_TOKEN cannot be used with an http:// STORAGE_URL: the token would
cross the network in cleartext. Use https:// or libsql://, or drop the token
for an unauthenticated local endpoint
```

`http://` with no token still validates — that is the local `sqld` case the scheme is allowed for, and it carries no credential to leak.

Worth separating this from the adjacent rule I declined earlier in this PR: requiring a token on `https://`. That one would reject a self-hosted `sqld` behind TLS with auth off, which is a legitimate deployment. This one rejects a combination that is never correct, so it is a hard error rather than a doc note.

Four tests in `env.schema.spec.ts` cover the matrix (`libsql://` needs a token, `https://` + token passes, `http://` alone passes, `http://` + token fails). Mutation-tested: reverting the schema fails the last one. Documented in `.env.example` and the configuration reference.

**Review by KIvanow — APPROVED** [review]
